server/src/server.ts

```diff
@@ -1,314 +1,342 @@
-			socket.on('close', () => delete this.clients[client.id])
-		})
 
-		this.server.on('error', (err: Error) => {
-			console.error('[TcpServer] Error:', err)
-			this.server.close()
-		})
+	public readonly server: net.Server;
+	public readonly clients: Map<number, TcpClient> = new Map();
+	private nextClientId: number = 0;
 
+	constructor(
+		public readonly handler: ProtocolHandler,
+	) {
+		this.server = net.createServer({}, (socket: net.Socket) => {
+			const client = new TcpClient(++this.nextClientId, socket, this, this.handler);
+			this.clients.set(client.id, client);
+			socket.on("end", () => this.clients.delete(client.id));
+			socket.on("close", () => this.clients.delete(client.id));
+		});
+		this.server.on("error", (err: Error) => {
+			console.error("[TcpServer] Error:", err);
+			this.server.close();
+		});
+		const { PORT = "12312", HOST = "127.0.0.1" } = process.env;
 		this.server.listen({ port: PORT, hostname: HOST }, () => {
-			console.log('[TcpServer] Listening on', HOST, PORT)
-		})
+			console.log("[TcpServer] Listening on " + HOST + ":" + PORT);
+		});
 	}
 
-	decrypt(buf: Buffer) {
-		return crypto.privateDecrypt(
-			{
-				key: this.keyPair.privateKey,
-				padding: crypto.constants.RSA_PKCS1_PADDING,
-			},
-			buf,
-		)
-	}
 }
 
-let nextClientId = 1
+// ============================================================ //
+// Crypto
+// ============================================================ //
 
-/** Prefixes packets with their length (UInt32BE);
- * handles Mojang authentication */
-export class TcpClient {
-	readonly id = nextClientId++
-	/** contains mojang name once logged in */
-	name = 'Client' + this.id
+const KEY_PAIR: KeyPairKeyObjectResult = crypto.generateKeyPairSync("rsa", { modulusLength: 1024 });
+const PUBLIC_KEY_BUFFER: Buffer = KEY_PAIR.publicKey.export({ type: "spki", format: "der" });
 
-	modVersion: string | undefined
-	gameAddress: string | undefined
-	uuid: string | undefined
-	mcName: string | undefined
-	world: string | undefined
+function decryptBuffer(buffer: Buffer): Buffer {
+	return crypto.privateDecrypt({
+		key: KEY_PAIR.privateKey,
+		padding: crypto.constants.RSA_PKCS1_PADDING
+	}, buffer);
+}
 
-	whitelisted: boolean | undefined
+// ============================================================ //
+// Clients
+// ============================================================ //
 
-	/** prevent Out of Memory when client sends a large packet */
-	maxFrameSize = 2 ** 24
+/**
+ * Prefixes packets with their length (UInt32BE) and handles Mojang-authentication
+ */
+export class TcpClient {
 
-	/** sent by client during handshake */
-	private claimedMojangName?: string
-	private verifyToken?: Buffer
-	/** we need to wait for the mojang auth response
-	 * before we can en/decrypt packets following the handshake */
+	// contains Mojang-name post-authentication
+	public name: string;
+	public modVersion?: string;
+	public gameAddress?: string;
```

Comment:
is this the server address?

Reply:
the minecraft server address, yes. I named it this way as to not confuse it with the sync server address on the client